questions/currying.md

```diff
@@ -0,0 +1,54 @@
+### How do you get both statements to evaluate to the same thing?
+
+```
+console.log(sum(2,3));   // Outputs 5
+console.log(sum(2)(3));  // Outputs 5
+```
+
+#### Answer
+
+This is an example of Currying, which is the process of breaking down a function into a series of functions that each take a single argument.
+
+In simplest terms, the second expression can be expressed as follows:
+
+```
+function sum (a) {
+  return function (b) {
+    return a + b;
+  }
+}
+
+sum(2)(3) // returns 5
+```
+
+Where the result of sum(2) is then passed into sum(3) and evaluated.
+
+However, this will not handle the first case of `sum(2,3)` unless we handle the potential second argument.
+
+In this case, we should use a slight conditional to handle both cases.
+
+```
+function sum (a, b) {
```

Comment:
There is a shorter way to write this

```js
function add(a,b) {
  const curry = (b) => a + b;
  return b ? curry(b) : curry;
}
```

Since multi line suggestions don't work too well I have to post it as a simple snippet, but you get the idea.

Reply:
You could also write a curry factory that would take a function and an arity

```js
const curry = (fn, arity=fn.length, context) => {
  const collector = args => ...rest => {
    args = args.concat(rest);
    return args.length >= arity ? fn.apply(context, args) : collector(args);
  }
  return collector([]);
}
```

Reply:
@skatcat31 Honestly, I find the factory example a bit overcomplicated for the needs of this question.

Reply:
Agreed. I think the shorter version offers more clarity for people unfamiliar with factories.

My thought was to use 'standard' functions (opposed to arrow functions) for clarity.

Reply:
@skatcat31 Do you want me to include it this PR, or would it make sense to be a separate question?

Reply:
It would be out of scope here as I severally doubt unless you're interviewing for a team lead position anyone would ask you to come up with an auto curry function